frontend_server/README.md

```diff
@@ -0,0 +1,35 @@
+Frontend Server
+==============
+
+Frontend server is simple wrapper around Dart Frontend. It is a Dart application
+that compiles Dart source into Dart Kernel binary(.dill-file).
+
+Frontend server runs in two modes:
+ - immediate mode, where Dart source file name is provided as command line
+ argument;
+ - interactive mode, where communication is happening over stdin/stdout.
+
+ Interactive mode instructions
+ ---
+- ```compile <input.dart>```
+ Compiles <input.dart> Dart source file with Dart Frontend, replies with ```result``` response.
+- ```recompile <boundary-key>```
+```<path/to/updated/file1.dart>```
+```<path/to/updated/file2.dart>```
+```...```
+```<boundary-key>```
+ Incrementally recompiles Dart program previously compiled in current session, taking into account changes in the listed files. Replies with ```result``` response.
+- ```accept```
+ Accepts results of incremental compilation, so that next recompilation request won't recompile previously reported changed files.
+- ```reject```
+ Rejects results of incremental compilation, so that next recompilation request will recompile previously reported changed files.
```

Comment:
Good question. The answer is somewhere in the middle:

We will not purposely recompile previously reported changed files, but we will be including their changes in the next result of the incremental compiler. Maybe a better phasing here would be to say:
> - ```accept```
> Accepts results of incremental compilation so that the next recompilation request doesn't include any libraries already loaded by the VM.
> - ```reject```
> Rejects results of incremental compilation, so the next recompilation request includes again the results of the previous compilation in addition to the new changes detected afterwards.

Basically internally the compiler tracks a few things:
 a. a hash of the library sources (when they were last compiled)
 b. the library kernel output associated with each library source file
 c. a hash of each kernel output library
 d. a table containing for each kernel library, the last hash that was accepted by the VM.

On step2 we update (a), (b), (c) for `mypart3.dart`.
On step3 we keep (d) unchanged: we decide that we will not use the new hash for `mypart3.dart` in (d) to indicate that this code was not received by the vm.
On step4 we don't recompile `mypart3.dart` (unless it was affected by the changes to the other files), but when we assemble what to send to the VM we notice that (d) has an older hash than the hash in (c) for mypart3.dart, so we'll send the library again.